### DISCUSSION TITLE
Swift UI views are not updated due to capturing viewStore (vs current state) in closures.

### DISCUSSION THREAD

**Discussion by grigorye** [pointfreeco/swift-composable-architecture#1012] [Q&A] [answered]
**Describe the bug**

A typical example of TCA usage advocates accessing state via `viewStore`, directly in bodies, like below (please note, `viewStore.send` is there just for completeness of the example, and has nothing to do to the problem per se - what is important is `viewStore.state`):

```
WithViewStore(store) { viewStore in
    WrapperX {
        ComponentY(viewStore.state, action: { viewStore.send(.action) })
    }
}
```

However, it results in no updates in Swift UI hierarchy in some "pretty innocent" cases of `WrapperX` implementation, based on closures.

**To Reproduce**

Please see https://github.com/grigorye/SwiftUIBodies.

**Expected behavior**

I would expect that `WithViewStore` would trigger updates inside WrapperX, in any variant of the implementation.

**Screenshots**

[Please see https://github.com/grigorye/SwiftUIBodies.](https://github.com/grigorye/SwiftUIBodies/blob/main/SwiftUIBodiesRun.png)

**Environment**
  - swift-composable-architecture version 0.33.1
  - Xcode 13.2.1
  - Swift 5.5
  - OS: Any

**Additional context**

The problem lies in the fact that `viewStore` is captured by reference (vs value, hence not accounting the value of state, if any), and the fact that body-generating closures in the wrappers are apparently part of the equation used when SwiftUI determines the need for re-render.

In the simplest case, a wrapper like below might be used to avoid the possibility of capturing the `viewStore`, but I'm not sure if it's applicable to more complex scenarios:

```
// Convenience initializer that makes it impossible to *accidentally* capture viewStore in a closure.
// See https://github.com/grigorye/SwiftUIBodies for more info.
extension WithViewStore where State: Equatable, Content: View {
    init(
        _ store: Store<State, Action>,
        content: @escaping (_ state: State, _ sendAction: @escaping (Action) -> Void) -> Content
    ) {
        self.init(store, content: { (viewStore: ViewStore<State, Action>) -> Content in
            content(viewStore.state, viewStore.send)
        })
    }
}
```

Usage:
```
WithViewStore(store) { state, sendAction in
    WrapperX {
        ComponentY(state, action: sendAction(.action))
    }
}
```

**Comment by tgrapperon** [comment]
I didn't have the time to investigate much, but please note that manually installing an `@ObservedObject viewStore: ViewStore` in your `Component` view instead of using `WithViewStore` fixes the issue. 

There's a subtlety in SwiftUI's invalidation process that I'm missing because I don't get why the `WithViewStore` content doesn't get the same treatment as its enclosing view. There are known issues with `WithViewStore` and `GeometryReader` or `Steppers` for example (#235, #224), and they're maybe related to the same cause.

A quick glance at the `WithViewStore` implementation makes me wonder if the `Content` closure shouldn't be eagerly evaluated in the `init` instead of capturing it. The `GeometryReader` closure is escaping, so I wonder if this is not causing too many closures being captured instead of evaluated, and changes being unaccounted. The same kind of things seems to happen in your demo with your `TextView`s. Many `View`s in SwiftUI are eagerly evaluating their `@ViewBuilder` arguments.
I don't know if it'll fix the issue and how much it will impact performance (we don't have `SwiftUI` benchmarks to check yet). I'll experiment later.

**Reply by tgrapperon** [reply]
After a quick try, making `WithViewStore` evaluate its closure eagerly in its `init` doesn't help (it even breaks everything).

I still don't exacly get wonder why something like
```swift
struct V: View {
  @ObservedObject viewStore: ViewStore
  var body: some View {
    content(viewStore)
  }
}
```
works whereas
```swift
struct V: View {
  var body: some View {
    WithViewStore(store) {
      content($0)
    }
  }
}
```
doesn't.

But I agree with @​mbrandonw's comment below that `()->ContentView` closures in initializers should be evaluated eagerly by default, especially for argument-less closures defining leaf views. This is something I started to do recently [after asking on twitter](https://twitter.com/tgrapperon/status/1484228600534609920?s=20&t=ecIQsAyLX4PUuF_3c1RTUw), and someone made the same remark about the vast majority of built-in init with closures being non-escaping (and a few inits are even inlined, so we can even explicitly spot the evaluation in SwiftUI's interface file!).

I have a very consequent project where I have many captured closures and where I'm not experiencing such issues, but that's most likely because I'm frequently using explicit `@ObservedObject var viewStore` instead of `WithViewStore` (I have a template I'm running for new features, so it comes by default).

**Reply by stephencelis** [reply]
@tgrapperon I explored this a bit further with a reduced example: https://gist.github.com/stephencelis/ea9255e5a3fe36149f76484baf9bf139

It appears that escaping closures that are passed reference types (_or_ even structs that simply wrap a reference type) are problematic, but escaping closures that are passed structs that also contain ~~some other data~~ _a copy of the changed state_ work fine.

If you swap out the `StoreWrapper` implementation for the version that is commented-out (which no longer holds directly onto a copy of store state), things break.

This "fix" isn't great, though. If we were to do the same in TCA it would mean introducing a new `ViewStoreWrapper` struct that is fed to `WithViewStore`, which is a breaking change, a whole new library concept that only exists to work around a SwiftUI bug, and has to duplicate all the surface area of `ViewStore` to get close to maintaining source compatibility.

Every attempt to tweak the above example to hide this extra type unfortunately failed:

  - I tried to expose a `WithViewStore` interface with an initializer that takes `(ViewStore<State, Action>) -> Content`, but storage that holds a `StoreWrapper` struct, and then we do an internal dance to convert the closure in the initializer. 
  - I tried to make an internal `ContentWrapper: View` that is initialized in `WithViewStore.body` to see if shuffling the data around would help...maybe there's something here, but the things I tried didn't really work.
  - I tried to make `StoreWrapper` a property wrapper with a projected value/initializer (like `Binding`) thinking that maybe omitting the `$` from the variable name would prefer the unwrapped value.

    ```swift
    WithViewStore(store) { viewStore in  // A `ViewStore`
                                         // vs.
    WithViewStore(store) { $viewStore in // A `ViewStoreWrapper`
    ```

    This doesn't work, but even if it did it's a bummer to expose this type publicly.

Just wanted to share my findings to see if you had any other ideas.

**Reply by grigorye** [reply]
> It appears that escaping closures that are passed reference types (or even structs that simply wrap a reference type) are problematic, but escaping closures that are passed structs that also contain some other data work fine.

Just to clarify this: if I get it correctly, "escaping closures that are passed structs that also contain some other data work fine" only if that "some other data" is changed with the state, correct? I'm reproducing it there in SwiftUIBodies with [now added TaggedGenerator](https://github.com/grigorye/SwiftUIBodies/commit/7ed70434592c144b9b4b5bc4999d6ca174ef5936) in the examples. Btw, what is interesting there is that with such struct, as opposed to other failing examples, it still *triggers* a re-rendering even when struct value is not changed, but it does that only once - probably some Swift UI optimization is catching up or whatever (see "XX" in front of `PlainText({ taggedGeneratorWith0.message() })` [there](https://github.com/grigorye/SwiftUIBodies/blob/main/SwiftUIBodiesRun.png)).

**Reply by stephencelis** [reply]
> only if that "some other data" is changed with the state, correct?

That's my understanding, yup!

**Reply by iampatbrown** [reply]
@stephencelis How about something like this?

```swift
public final class ViewStore<State, Action>: ObservableObject {
  ...
  private init(viewStore: ViewStore<State, Action>) {
    self._send = viewStore._send
    self._state = viewStore._state
    self.objectWillChange = viewStore.objectWillChange
  }

  func copy() -> Self {
    Self(viewStore: self)
  }
}

public struct WithViewStore<State, Action, Content> {
  ...
  private var _body: Content {
    ...
    return self.content(self.viewStore.copy())
  }
}
```

It's not exactly a copy of the `ViewStore` but it forces the identity to change. This should mean escaping closures treat it as an updated value.

**Reply by tgrapperon** [reply]
@stephencelis 
>  that only exists to work around a SwiftUI bug.

I think the issue is broader than that, and I'm starting to wonder if it's not the expected behavior from the very beginning. We experience it with `GeometryReader` because this is the most common example of a view with some deferred content, but if you try to replace `GeometryReader` with a `WithViewStore` from an unrelated store, you get the same behavior:
```swift
var body: some View {
  WithViewStore(self.store) { viewStore in
    Text("\(viewStore.state)") // This updates
    WithViewStore(anotherStore) { _ in
    // Things that are involving `viewStore` aren't updated anymore from there
…
```

I think I'm getting what's going on, but I can easily confuse myself, and the evidences we have about how it works are sparse, so my conclusions can be completely off. Here's what I know or assume to be true (mostly from my recollection of WWDC's [Demystify SwiftUI](https://developer.apple.com/videos/play/wwdc2021/10022/) that I should probably watch again):
- SwiftUI `View`'s have an identity and dependencies. I don't think the identity is relevant here. It's simply the path of a given view in the view tree. Dependencies are super interesting though. It roughly corresponds to the stored properties of each view. A graph of dependencies is built along the view tree. When a dependency is invalidated, the framework computes the orbit for this dependency and invalidates the related views
- SwiftUI is very conservative, and only invalidates the orbit of the dependencies that changed.

_[EDIT: I have a very long explanation about what I think is happening, but I ditched it, as it was mostly suppositions]_
I don't grasp exactly what's happening, but it looks like `SwiftUI` decides that if `viewStore` has the same memory representation, it's not worth re-evaluating the `body`. I don't know how it happens yet (nor if it does). If the `viewStore` value changes ( because you save the `state`, an `UUID` or if the `store` instance changes (or if you change it artificially as @iampatbrown suggested)), the body gets evaluated. This seems related to the way SwiftUI constructs the orbit of a dependency, but this is a black box. I don't know for example if the dependency graph is based on the declarative side (the `body`'s) or on the "system view" side. I don't know if this is realizable in Swift, but it's behaving like it supposes that if the capture list of a closure is the same (in terms of memory), the closure result will be the same.

So it seems indeed to be a broader issue than mixing `GeometryReader` or `ScrollViewReader`. It looks like it can happen as soon as we're handling `View`s with lazy content evaluation. Maybe we don't see this problem often in TCA  because `Store` and `ViewStore` are not reused, and `ViewStore` instances are thus changing very often and closures are getting reevaluated. But it's only a guess, and I can't explain yet why it works when `@ObservedObject` is stored as a view's dependency. There are probably several layers of optimization in the way.

Maybe it's not worth addressing the issue for the current flavor of TCA, as `ViewStore` is a very fundamental type in the library. A potential solution would maybe to have the `ViewStore` argument of `WithViewStore` be a value type (like `GeometryProxy`), like you did with `ViewStoreWrapper`. In this case, we would indeed handle a pure value type that would be different when the state is different in `body`[^1]. I think that if some major refactor happens in the library (like the `proto` branch discussed in #832), this could be something to consider for all `ViewStore`s, and not only the ones used in `WithViewStore`. Off the top of my head, `ViewStore` is only a class to be `ObsersableObject`. This can probably be circumvented with custom a `DynamicProperty` wrapper to store the `Store`. But I agree that having this only for `WithViewStore` adds a degree of indirection that feels unnecessary.

[^1]: In the experimental PR #935 where I aimed to reuse `Store` and `ViewStore` instances, I suggested renaming the `ViewStore`  as `StoreView`, as `Store` could inherit a `view: StoreView` property that would provide a very light "view" into (and to) the store. This would have worked in conjunction with some dedicated property wrappers implementing `DynamicProperty`.

**Reply by iampatbrown** [reply]
> I can't explain yet why it works when @ObservedObject is stored as a view's dependency.

I believe the difference is that you are passing the view to the closure and not the view store. For example:

```swift
class Counter: ObservableObject {
  @Published var count = 0
}

struct Wrapper<Content: View>: View {
  @ViewBuilder var content: () -> Content
  var body: some View { self.content() }
}

struct ContentView: View {
  @ObservedObject var counter = Counter()

  var body: some View {
    Wrapper {
      Stepper(onIncrement: { counter.count += 1 }, onDecrement: { counter.count -= 1 }) {
        Text("\(counter.count)")
      }
    }
  }
}
```

could be written as:

```swift
struct ContentView: View {
  @ObservedObject var counter = Counter()

  var body: some View {
    Wrapper { [view = self] in
      Stepper(onIncrement: { view.counter.count += 1 }, onDecrement: { view.counter.count -= 1 }) {
        Text("\(view.counter.count)")
      }
    }
  }
}
```
`WithViewStore` is more equivalent to:

```swift
struct ContentView: View {
  @ObservedObject var counter = Counter()

  var body: some View {
    Wrapper { [counter = self.counter] in
      Stepper(onIncrement: { counter.count += 1 }, onDecrement: { counter.count -= 1 }) {
        Text("\(counter.count)")
      }
    }
  }
}
```

explicitly capturing the observed object's wrapped value causes the same issue.

I think creating a copy of the ViewStore is a fairly lightweight way to fix this from a `WithViewStore` perspective.

**Reply by tgrapperon** [reply]
Thanks @iampatbrown! That makes sense I guess.

> I think creating a copy of the ViewStore is a fairly lightweight way to fix this from a WithViewStore perspective.

That's indeed transparent for the user. I have two small reservations:
- Wouldn't this cause over-invalidation of the dependency graph? We're always providing a new value when building the view, even if the `viewStore`'s state hasn't changed. Maybe we can perform additional checks somewhere to vend a new copy only when needed (we can maybe lazily instantiate the copy on access, store it in the `viewStore` itself to reuse it, and delete it on `objectWillChange`?). We can store the copy in the `ViewStore` or maybe in `WithViewStore`, into some `@State` box. But maybe this is not needed. I need to think more about it. As I'm getting it, the copy should happen in the `init` instead of `body` to be detected, isn't it?
- ~~Reassigning the `objectWillChange` publisher is a little scary at first sight: is it the only thing we need to perform to make both viewStores identical as `ObservableObject`, are we not breaking some internal wiring with `@Published` properties for example?~~ _EDIT: I've checked again your implementation, and you initialize the `ViewStore` directly from the original, so I have no concern over this._

I still think that having a value type + dedicated property wrapper could be desirable in the long run, but using copies is indeed a lightweight fix to the nesting issue.

**Reply by tgrapperon** [reply]
@iampatbrown, after thinking more about it, I don't have reservations anymore about over-invalidation. The `WithViewStore` body will run only when the state has changed, so it's not problematic to use a new instance when doing this, and the `viewStore` will behave like a value type, with a new value when the state changes.
And I don't have issues either with the way you build a copy as it happens in the `ViewStore`'s initializer. The `@ObservedObject` instance is private, so it changes nothing for the user, and the wiring makes them talk to it directly, without much indirection.

So this seems a very good solution, optimal in every aspect! 
Do you want to prepare a PR? or I can give it a try if you don't have the time.

**Reply by iampatbrown** [reply]
> The WithViewStore body will run only when the state has changed, so it's not problematic to use a new instance when doing this, and the viewStore will behave like a value type, with a new value when the state changes.

@tgrapperon This is basically what I was thinking. I wanted to ensure my understanding was correct before getting back to you though :)

I'm happy to do a PR, I'd like to wait for some input from @stephencelis and @mbrandonw though. There could be some stylistic considerations as well as implications extending beyond this particular issue. 

If you feel like putting together a PR now though, go for it :) We can use that for further discussion.

**Reply by tgrapperon** [reply]
@iampatbrown I've drafted #1015 in this spirit.
It seems to fix the issue by the way!

**Reply by iampatbrown** [reply]
Awesome. Good work!

**Comment by mbrandonw** [comment, marked as answer]
This isn't a problem with TCA per se, and would be easily reproducible in a vanilla SwiftUI application too. In fact we have some bugs filed with Apple that demonstrate the problem ([here](https://gist.github.com/mbrandonw/cc5da3d487bcf7c4f21c27019a440d18) and [here](https://gist.github.com/mbrandonw/dee2ceac2c316a1619cfdf1dc7945f66)).

However, there is an easy change that fixes the problem, and it's probably the better way of constructing your views anyway. Although most SwiftUI views involve closures in some way, 99% of the time they are non-escaping closures. That means the closures aren't actually meant to be stored away and lazily evaluated later, but rather they just serve as a convenient syntax and are evaluated immediately.

For example, the `Section` view takes view builder closures for parent, content and footer views, but none of those closures are escaping:

```swift
extension Section where Parent : View, Content : View, Footer : View {
  public init(@ViewBuilder content: () -> Content, @ViewBuilder header: () -> Parent, @ViewBuilder footer: () -> Footer)
}
```

This is how SwiftUI is meant to be used.

So, if you change all of your views to no longer take escaping closures in the initializers, and instead eager evaluate the closures, you will see that everything works as you expect:

```swift
struct FancyTextWrapper: View, Traceable {
  let content: Text
  init(@ViewBuilder content: () -> Text) {
    self.content = content()
  }
  var body: some View {
    content
  }
}
```

Hope it helps!

Also, I'm going to convert this issue to a discussion since it does not seem to be a bug in the library.

**Reply by grigorye** [reply]
Thanks! This looks really interesting - will try that. Yeah, I did not meant it's a bug in TCA per se, rather something that we encountered while working with TCA and Swift UI, that is still new for us as developers. Wondering if something like this (eager evaluation of closures) could be/is put in best practices or somewhere else - it does not look evident at all, from my perspective.

**Reply by grigorye** [reply]
Checked it and I can confirm that eager evaluation fixes the problem. All in all, now it looks reasonable to me: on a change of the state, we want to regenerate the body, right away, and the most straightforward way to achieve that is just evaluating all the closures responsible for the generation. Lazy evaluation provides an alternative (if it's sound at all) but, as we figured it out, it's risky, does not provide any benefits (including performance, as far as I could imagine), and is just a more complex solution for the same task.

**Comment by grigorye** [comment]
>bugs filed with Apple that demonstrate the problem ([here](https://gist.github.com/mbrandonw/cc5da3d487bcf7c4f21c27019a440d18) and [here](https://gist.github.com/mbrandonw/dee2ceac2c316a1619cfdf1dc7945f66)).

Just in case, I could not reproduce anything meaningful with the second example. Left a comment in the gist.
